File: code/86.分隔链表.go
/*
 * @lc app=leetcode.cn id=86 lang=golang
 *
 * [86] 分隔链表
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
	dummy := ListNode{
		Val:  0,
		Next: head,
	}

	pre := &dummy
	for pre.Next != nil && pre.Next.Val < x {
		pre = pre.Next
	}

	cur := pre
	for cur.Next != nil {
		if cur.Next.Val < x {
			next := cur.Next
			cur.Next = next.Next
			next.Next = pre.Next
			pre.Next = next
			pre = next
		} else {
			cur = cur.Next
		}
	}

	return dummy.Next
}

File: code/86.分隔链表.py
#
# @lc app=leetcode.cn id=86 lang=python
#
# [86] 分隔链表
#
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        pre = dummy
        while(pre.next != None and pre.next.val < x):
            pre = pre.next

        cur = pre
        while(cur.next != None):
            if(cur.next.val < x):
                next = cur.next
                cur.next = next.next
                next.next = pre.next
                pre.next = next
                pre = next
            else:
                cur = cur.next
        
        return dummy.next
        

